### ISSUE TITLE
Allocate the same resource to several arrivals without losing attribute information

### ISSUE THREAD

**Issue by MichiSchmid** [r-simmer/simmer.bricks#13]
I've been trying to implement a simmer-routine to bundle arrivals for one resource without losing attribute-information about the arrivals (`simmer::batch`). This might be an issue for simmer itself. 

Imagine one lift (`resource`) transporting people (`arrivals`) to different floors. Let's say each person uses the lift on their own, a simmer-implementation might look like this: 

```
library(simmer)

env <- simmer() %>%
  add_resource("lift")

t_lift <- trajectory() %>%
  set_attribute("to_floor", function() sample(1:25, 1)) %>%
  log_(function() paste("Going to floor:", get_attribute(env, "to_floor"))) %>%
  seize("lift", 1) %>%
  log_("Lift seized") %>%
  timeout(function() 2*get_attribute(env, "to_floor")) %>%
  release("lift", 1) %>%
  log_("Lift released")

env %>%
  add_generator("person", t_lift, at(15*seq(0,3))) %>%
  run()
```

which gives: 

```
0: person0: Going to floor: 22
0: person0: Lift seized
15: person1: Going to floor: 2
30: person2: Going to floor: 8
44: person0: Lift released
44: person1: Lift seized
45: person3: Going to floor: 1
48: person1: Lift released
48: person2: Lift seized
64: person2: Lift released
64: person3: Lift seized
66: person3: Lift released
```

------------------------------------------------------------------------------------

If several persons should use the same lift, I can't find a suitable solution. When batching `arrivals`, the attribute-information can't be retrieved anymore in order to define the timeout-time for the `resources`.  

```
env <- simmer() %>%
  add_resource("lift")

t_lift <- trajectory() %>%
  set_attribute("to_floor", function() sample(1:25, 1)) %>%
  log_(function() paste("Going to floor:", get_attribute(env, "to_floor"))) %>%
  batch(4) %>%
  seize("lift", 1) %>%
  log_("Lift seized") %>%

  timeout(function() {
    # some routine to define the travel time to the different floores.
    # For this we need the arrivals' attributes inside the batch...
    time_to_first_stop <- get_attribute(env, "to_floor")
    # ...
  }) %>%

  release("lift", 1) %>%
  log_("Lift released")

env %>%
  add_generator("person", t_lift, at(15*seq(0,3))) %>%
  run()
```
```
0: person0: Going to floor: 1
15: person1: Going to floor: 20
30: person2: Going to floor: 3
45: person3: Going to floor: 13
45: batch0: Lift seized
Error: 'batch0' at 45.00 in [Log]->Timeout->[Release]:
 missing value (NA or NaN returned)
```

To sum up, I see two questions with this Issue: 
- Is there a way to retrieve an arrivals' attribute-information from inside a batch? 
- Is there a way to allocate the same resource to several arrivals? (If there is a solution to this question, the first question becomes obsolete from my point of view). 

Best, Michael

**Comment by Enchufa2** [comment]
>
>  To sum up, I see two questions with this Issue:
>
>    - Is there a way to retrieve an arrivals' attribute-information from
>    inside a batch?
>
> No, there is not. Batches' attributes are propagated downwards to
individual arrivals, but not the other way around.

>
>    - Is there a way to allocate the same resource to several arrivals?
>    (If there is a solution to this question, the first question becomes
>    obsolete from my point of view).
>
> I don't think that a batch is an appropriate model for this use case,
because people behave independently (different origins and destinations) in
general.

>

**Comment by MichiSchmid** [comment]
> No, there is not. Batches' attributes are propagated downwards to individual arrivals, but not the other way around.

Ok, thanks for this info. 

> I don't think that a batch is an appropriate model for this use case, because people behave independently (different origins and destinations) in general.

Assume they all have the same origin (for example a construction site, where Items need to be transported from the groundfloor to different floors). In this case a batch would make sense I guess, since you want to use the lift for several items at a time. An allocation of one resource to several arrivals would be very helpful in this case. Do you see a possible way using simmer for this kind of problem?

**Comment by Enchufa2** [comment]
> Assume they all have the same origin (for example a construction site, where Items need to be transported from the groundfloor to different floors).

Still, if they are headed to different floors, you have to break the batch in the first stop. Just define your lift's capacity with the required number of items (see `?add_resource`).

**Comment by MichiSchmid** [comment]
I see your point. 
Let's say we set the lifts capacity to 4. This would lead to 4 individual lifts (L1 - L4 with one common queue) acting as independant systems. If I'm not mistaken, this would imply that e.g. L1 could be at a different floor than L2, L3, L4 which wouldn't be a very appropriate model of a Lift with capacity 4. 
Please correct me if that's not correct.

**Comment by Enchufa2** [comment]
There's a previous step. The floor itself is a resource (at least, a queue) that you need to seize before accessing an elevator. A lift with capacity 4 would be a single resource with capacity = 4 people. If you want 4 lifts, I would use 4 independent resources.

But anyway, this issue was more appropriate for `simmer` instead of `simmer.bricks` in the first place, and now it turned into a discussion about modelling. So I'm closing this and I encourage you to raise this in the [mailing list](https://groups.google.com/forum/#!forum/simmer-devel), where not only me, but other users may help and share their expertise.

**Comment by MichiSchmid** [comment]
Okay, thanks for your time.
